sjsonnet/src/sjsonnet/Std.scala

```diff
@@ -1383,18 +1347,9 @@ class Std {
         }
       } else if (!keyF.isInstanceOf[Val.False]) {
         val keyFFunc = keyF.asInstanceOf[Val.Func]
-
         val o1Key = keyFFunc.apply1(v, pos.noOffset)(ev)
         val o2Key = keyFFunc.apply1(out.last, pos.noOffset)(ev)
-        val o1KeyExpr = Materializer.toExpr(Materializer.apply(o1Key)(ev))(ev)
-        val o2KeyExpr = Materializer.toExpr(Materializer.apply(o2Key)(ev))(ev)
-
-        val comparisonExpr = Expr.BinaryOp(dummyPos, o1KeyExpr, BinaryOp.OP_!=, o2KeyExpr)
-        val exprResult = ev.visitExpr(comparisonExpr)(ValScope.empty)
-
-        val res = Materializer.apply(exprResult)(ev).asInstanceOf[ujson.Bool]
-
-        if (res.value) {
+        if (!ev.equal(o1Key, o2Key)) {
```

Comment:
For my own intuition:

Why isn't this

```
if (!ev.equal(o1Key.force, o2Key.force))
```

?

I don't fully understand why we call `.force` in the other equality comparisons in other contexts so I'm unsure if we should be doing it here.

Reply:
My understanding is that there are 2 ways of doing it in sjsonnet:

- You can materialize it to json with Materializer and then evaluate it
- You can for the evaluation of a Val and then compare it.

For consistency, I'm picking the latter.

Reply:
Yep, 100% on the same page about avoiding `Materializer` usage.

But what I'm asking here is about whether we force the evaluation of the key expression vs. comparing lazy, e.g. why this is

```
if (!ev.equal(o1Key, o2Key))
```

and not

```
if (!ev.equal(o1Key.force, o2Key.force))
```

My understanding is that `ev.equal` has some fastpaths to allow for comparison of values without actually computing them, e.g. if we have a lazy not-yet-evaluated `x` then we can sometimes determine that x == x` via reference equality.

I noticed that some of the `ev.equal` call sites call `.force` in order to guarantee that the values are computed and I'm wondering why we do that. Does it have something to do with formal semantics of when evaluation occurs? I don't think it's necessary for correctness w.r.t. the final answer, but it might affect the side-effecting behavior of whether evaluation is performed and whether assertions in the to-be-computed subtree are evaluated?

Reply:
ah ok.
unclear as to why ev.equal does not take Lazy.
in the end, it does not matter, because right now, we force the evaluation of all parameters (not inside the arrays though) in Var.Builtin()

```
    final def evalRhs(scope: ValScope, ev: EvalScope, fs: FileScope, pos: Position): Val = {
      val args = new Array[Val](params.names.length)
      var i = 0
      var j = scope.length - args.length
      while(i < args.length) {
        args(i) = scope.bindings(i).force
        i += 1
      }
      evalRhs(args, ev, pos)
    }
```